src/components/data/viewers/TextViewer.js

```diff
@@ -52,66 +33,171 @@ export default function TextViewer(props) {
+                );
+                if (fileTypeAvu && fileTypeAvu.length > 0) {
+                    setDetectedMode(fileTypeAvu[0].value);
+                }
+            },
+            onError: (error) =>
+                console.log("Unable to fetch metadata for viewer. " + error), // fail silently.
+        },
+    });
+
+    const [setDiskResourceMetadata] = useMutation(setFilesystemMetadata, {
+        onSuccess: (resp) => {
+            console.log(resp); // background optional call. No need to notify user.
+        },
+        onError: (error) => {
+            console.log(error); // fail silently.
+        },
+    });
+
+    const getContent = () => {
+        return editorValue;
+    };
+
+    const updateNewFileMetadata = (details) => {
+        const metadata = {
+            avus: [
+                {
+                    attr: viewerConstants.IPC_VIEWER_TYPE,
+                    value: createFileType,
+                    unit: "",
+                },
+            ],
+            "irods-avus": [],
+        };
+
+        setDiskResourceMetadata({
+            dataId: details?.file.id,
+            metadata,
+        });
+    };
+
+    const busy = loading || isFileSaving || isFetchingMetadata;
 
     return (
-        <PageWrapper
-            appBarHeight={120}
-            id={build(baseId, ids.VIEWER_PLAIN, fileName)}
-        >
+        <>
             <Toolbar
                 baseId={build(baseId, ids.VIEWER_DOC, ids.TOOLBAR)}
                 path={path}
                 resourceId={resourceId}
                 allowLineNumbers={true}
                 showLineNumbers={showLineNumbers}
-                onShowLineNumbers={(show) => setShowLineNumbers(show)}
+                onShowLineNumbers={setShowLineNumbers}
+                wrapText={wrapText}
+                onWrapText={setWrapText}
                 handlePathChange={handlePathChange}
                 onRefresh={onRefresh}
                 fileName={fileName}
-                modeSelect={
-                    <ModeSelect
-                        baseId={build(
-                            baseId,
-                            ids.VIEWER_PLAIN,
-                            ids.SELECT_MODE_INPUT
-                        )}
-                        style={{ width: 200 }}
-                        mode={mode}
-                        handleModeSelect={handleModeSelect}
-                    />
-                }
+                editable={editable}
+                dirty={dirty}
+                createFileType={createFileType}
+                onNewFileSaved={onNewFileSaved}
+                getFileContent={getContent}
+                onSaving={() => setFileSaving(true)}
+                onSaveComplete={(details) => {
+                    setFileSaving(false);
+                    setDirty(false);
+                    if (createFileType) {
+                        updateNewFileMetadata(details);
+                    }
```

Comment:
Right. I was looking for a `PATCH` endpoint and couldn't find one. Is there a end-point to update metadata instead of setting ?

Reply:
The [POST /terrain/secured/filesystem/{data-id}/metadata endpoint](https://de.cyverse.org/terrain/docs/index.html#!/default/post_terrain_secured_filesystem_data_id_metadata) will update an AVU if the AVU's `id` is included in the request, otherwise it will add that AVU to the metadata.

Reply:
`setFilesystemMetadata` is indeed calling that end-point from serviceFacade.

Reply:
Oh, sorry. I was getting the POST/PUT endpoints in apps/metadata confused with the data-info endpoints.

It appears data-info is using `POST` to set AUVs with the metadata service's `PUT` endpoint, but it uses a `PATCH` endpoint to add/update AVUs with the metadata service's `POST` endpoint; however, terrain is only exposing data-info's `POST` endpoint that sets AVUs 😞